explainers/beforematch-spec-draft.md

```diff
@@ -0,0 +1,51 @@
+# beforematch event
+
+_These changes will likely go into the
+[find-in-page](https://html.spec.whatwg.org/multipage/interaction.html#find-in-page)
+section of the HTML spec_
+
+When a new active match is set, either by advancing through the match list or
+due to a new find-in-page request, a `beforematch` event is fired on a node
+identified by the active match [range
+start](https://dom.spec.whatwg.org/#concept-range-start). This process follows
+the following algorithm:
+
+1. Identify the match _M_ which will become the new active match. _M_'s DOM
+range must not be [collapsed](https://dom.spec.whatwg.org/#range-collapsed).
+
+2. At the next rendering opportunity, specifically step 12 of [Update the
+rendering](https://html.spec.whatwg.org/#rendering-opportunity) step:
+
+      2.1. If the range that constitutes match _M_ is
+        [collapsed](https://dom.spec.whatwg.org/#range-collapsed), restart the
+        algorithm at Step 1.
+
+      2.2. Fire a `beforematch` event on an element identified by _M_'s range start node.
+
+3. If the event fired in Step 2.2 was handled by any beforematch handler, then
+delay Step 4 until the next rendering opportunity (the next Step 12 of [Update the
+rendering](https://html.spec.whatwg.org/#rendering-opportunity)). Otherwise,
+proceed to Step 4 immediately.
```

Comment:
I haven't implemented this step or planned on doing so.

I suppose I could give it a shot, but it would require carefully traversing the event bubbling route that the event will take up the DOM tree and checking for beforematch handlers all the way.

Do you think this is important to do? I suppose it would make find-in-page one frame faster in most normal cases which don't involve beforematch, but I *personally* haven't noticed any slowness from the double rAF I put in m86.

Reply:
Do you mean the "otherwise" part? I think it's ok to unconditionally delay.

Reply:
Yes, we are unconditionally delaying right now, and it would be a little complicated to check if we actually called a beforematch handler due to event bubbling.

Reply:
Sounds good, I will remove the condition.